backend/x86_binary_emitter.ml

```diff
@@ -0,0 +1,1764 @@
+  | ((Reg32 _ | Mem _ | Mem64_RIP _) as rm), Regf reg ->
+    buf_int8 b 0x66;
+    emit_mod_rm_reg b no_rex [ 0x0F; 0x7E ] rm (rd_of_regf reg)
+  | _ -> assert false
+
+let emit_movq b ~dst ~src =
+  (* It seems there is a choice to make on how we encode instructions here
+     as there are different encoding possible for the same operation.
+     See https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-instruction-set-reference-manual-325383.pdf,
+     pages 707 and 755.  *)
+  match (dst, src) with
+  | Regf reg, ((Reg64 _ | Mem _ | Mem64_RIP _) as rm) ->
+    buf_int8 b 0x66;
+    emit_mod_rm_reg b rexw [ 0x0F; 0x6E ] rm (rd_of_regf reg)
+  | ((Reg64 _ | Mem _ | Mem64_RIP _) as rm), Regf reg ->
+    buf_int8 b 0x66;
+    emit_mod_rm_reg b rexw [ 0x0F; 0x7E ] rm (rd_of_regf reg)
+  | Regf reg, ((Regf _) as rm) ->
+    buf_int8 b 0xF3;
+    emit_mod_rm_reg b no_rex [ 0x0F; 0x7E ] rm (rd_of_regf reg)
+  | _ -> assert false
+
+let emit_movsd b dst src =
+  match (dst, src) with
+  | Regf reg, ((Regf _ | Mem _ | Mem64_RIP _) as rm) ->
+      buf_int8 b 0xF2;
+      emit_mod_rm_reg b 0 [ 0x0f; 0x10 ] rm (rd_of_regf reg)
+  | ((Mem _ | Mem64_RIP _) as rm), Regf reg ->
+      buf_int8 b 0xF2;
+      emit_mod_rm_reg b 0 [ 0x0f; 0x11 ] rm (rd_of_regf reg)
+  | _ ->
+      Format.eprintf "src=%a dst=%a@." print_old_arg src print_old_arg dst;
+      assert false
+
+let emit_movss b dst src =
+  match (dst, src) with
+  | Regf reg, ((Regf _ | Mem _ | Mem64_RIP _) as rm) ->
+      buf_int8 b 0xF3;
+      emit_mod_rm_reg b 0 [ 0x0f; 0x10 ] rm (rd_of_regf reg)
+  | ((Mem _ | Mem64_RIP _) as rm), Regf reg ->
+      buf_int8 b 0xF3;
+      emit_mod_rm_reg b 0 [ 0x0f; 0x11 ] rm (rd_of_regf reg)
+  | _ -> assert false
+
+let emit_andpd b dst src =
+  match (dst, src) with
+  | Regf reg, ((Regf _ | Mem _ | Mem64_RIP _) as rm) ->
+      buf_int8 b 0x66;
+      emit_mod_rm_reg b 0 [ 0x0f; 0x54 ] rm (rd_of_regf reg)
+  | _ -> assert false
+
+let emit_bsf b ~dst ~src =
+  match (dst, src) with
+  | Reg16 reg, ((Reg16 _ | Mem _ | Mem64_RIP _) as rm)
+  | Reg32 reg, ((Reg32 _ | Mem _ | Mem64_RIP _) as rm) ->
+    (* BSF r16, r/m16 and BSF r32, r/m32 *)
+    emit_mod_rm_reg b 0 [ 0x0F; 0xBC ] rm (rd_of_reg64 reg)
+  | Reg64 reg, ((Reg64 _ | Mem _ | Mem64_RIP _) as rm) ->
+    (* BSF r64, r/m64 *)
+    emit_mod_rm_reg b rexw [ 0x0F; 0xBC ] rm (rd_of_reg64 reg)
+  | _ -> assert false
+
+let emit_bsr b ~dst ~src =
+  match (dst, src) with
+  | Reg16 reg, ((Reg16 _ | Mem _ | Mem64_RIP _) as rm)
+  | Reg32 reg, ((Reg32 _ | Mem _ | Mem64_RIP _) as rm) ->
+    (* BSR r16, r/m16 and BSR r32, r/m32 *)
+    emit_mod_rm_reg b 0 [ 0x0F; 0xBD ] rm (rd_of_reg64 reg)
+  | Reg64 reg, ((Reg64 _ | Mem _ | Mem64_RIP _) as rm) ->
+    (* BSR r64, r/m64 *)
+    emit_mod_rm_reg b rexw [ 0x0F; 0xBD ] rm (rd_of_reg64 reg)
+  | _ -> assert false
+
+let imm8_of_rounding rounding =
```

Comment:
This could probably be shared with [this type](https://github.com/ocaml-flambda/flambda-backend/blob/main/backend/amd64/arch.ml#L76)
and/or [this conversion function](https://github.com/ocaml-flambda/flambda-backend/blob/main/backend/x86_proc.ml#L244).